File: e2e/state_operations_test.go
//go:build e2e

package e2e

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

// TestPulumiStackExport runs `pulumi stack export` on a stack that has been deployed to.
func TestPulumiStackExport(t *testing.T) {
	truncateDB(t)
	env := newTestEnv(t, "export-project")
	env.login()

	fqn := devOrgLogin + "/export-project/dev"
	env.run("stack", "init", "--stack", fqn)
	env.run("up", "--yes", "--stack", fqn)

	stdout, _ := env.run("stack", "export", "--stack", fqn)

	var deployment map[string]json.RawMessage
	if err := json.Unmarshal([]byte(stdout), &deployment); err != nil {
		t.Fatalf("export output is not valid JSON: %v\noutput: %s", err, stdout)
	}

	if _, ok := deployment["version"]; !ok {
		t.Fatal("expected 'version' field in export output")
	}
	if _, ok := deployment["deployment"]; !ok {
		t.Fatal("expected 'deployment' field in export output")
	}
}

// TestPulumiStackExportImport runs export then import — the round-trip.
func TestPulumiStackExportImport(t *testing.T) {
	truncateDB(t)

	home := t.TempDir()
	projectDir := t.TempDir()
	pulumiYAML := `name: exportimport-project
runtime: yaml
description: E2E test export/import
outputs:
  greeting: "hello from export-import"
`
	if err := os.WriteFile(filepath.Join(projectDir, "Pulumi.yaml"), []byte(pulumiYAML), 0o644); err != nil {
		t.Fatalf("write Pulumi.yaml: %v", err)
	}

	env := &testEnv{
		t:           t,
		home:        home,
		projectDir:  projectDir,
		accessToken: devAuthToken,
	}
	env.login()

	fqn := devOrgLogin + "/exportimport-project/dev"
	env.run("stack", "init", "--stack", fqn)
	env.run("up", "--yes", "--stack", fqn)

	// Export the state.
	exportStdout, _ := env.run("stack", "export", "--stack", fqn)

	// Write export to a temp file for import.
	exportFile := filepath.Join(t.TempDir(), "state.json")
	if err := os.WriteFile(exportFile, []byte(exportStdout), 0o644); err != nil {
		t.Fatalf("write export file: %v", err)
	}

	// Import the state back.
	stdout, stderr := env.run("stack", "import", "--stack", fqn, "--file", exportFile)
	output := combinedOutput(stdout, stderr)
	t.Logf("import output: %s", output)

	// Verify the stack output is still correct after import.
	outStdout, _ := env.run("stack", "output", "--json", "--stack", fqn)
	if !strings.Contains(outStdout, "hello from export-import") {
		t.Fatalf("expected 'hello from export-import' in stack output after import, got: %s", outStdout)
	}

	// Verify a new import update was created in the DB.
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	var importCount int
	err := dbPool.QueryRow(ctx, `SELECT COUNT(*) FROM updates WHERE kind = 'import' AND status = 'succeeded'`).Scan(&importCount)
	if err != nil {
		t.Fatalf("query import updates: %v", err)
	}
	if importCount == 0 {
		t.Fatal("expected at least 1 import update record")
	}
}

// TestPulumiStackExportEmpty tests exporting an empty stack (no deployments yet).
func TestPulumiStackExportEmpty(t *testing.T) {
	truncateDB(t)
	env := newTestEnv(t, "export-empty-project")
	env.login()

	fqn := devOrgLogin + "/export-empty-project/dev"
	env.run("stack", "init", "--stack", fqn)

	stdout, _ := env.run("stack", "export", "--stack", fqn)

	var deployment map[string]json.RawMessage
	if err := json.Unmarshal([]byte(stdout), &deployment); err != nil {
		t.Fatalf("export output is not valid JSON: %v\noutput: %s", err, stdout)
	}

	if _, ok := deployment["version"]; !ok {
		t.Fatal("expected 'version' field in empty export output")
	}
}

// TestPulumiStackImportToNewStack imports state into a fresh stack with no prior deploys.
func TestPulumiStackImportToNewStack(t *testing.T) {
	truncateDB(t)

	home := t.TempDir()
	projectDir := t.TempDir()
	pulumiYAML := `name: import-fresh-project
runtime: yaml
description: E2E test import to fresh stack
outputs:
  msg: "imported state"
`
	if err := os.WriteFile(filepath.Join(projectDir, "Pulumi.yaml"), []byte(pulumiYAML), 0o644); err != nil {
		t.Fatalf("write Pulumi.yaml: %v", err)
	}

	env := &testEnv{
		t:           t,
		home:        home,
		projectDir:  projectDir,
		accessToken: devAuthToken,
	}
	env.login()

	// Create two stacks — deploy to one, export, import to another.
	srcFQN := devOrgLogin + "/import-fresh-project/src"
	dstFQN := devOrgLogin + "/import-fresh-project/dst"
	env.run("stack", "init", "--stack", srcFQN)
	env.run("stack", "init", "--stack", dstFQN)

	env.run("up", "--yes", "--stack", srcFQN)

	exportStdout, _ := env.run("stack", "export", "--stack", srcFQN)

	exportFile := filepath.Join(t.TempDir(), "state.json")
	if err := os.WriteFile(exportFile, []byte(exportStdout), 0o644); err != nil {
		t.Fatalf("write export file: %v", err)
	}

	// Import into the fresh (never deployed) stack. --force needed because stack names differ.
	stdout, stderr := env.run("stack", "import", "--stack", dstFQN, "--file", exportFile, "--force")
	t.Logf("import to fresh stack: %s", combinedOutput(stdout, stderr))

	// Verify exported output matches.
	outStdout, _ := env.run("stack", "output", "--json", "--stack", dstFQN)
	if !strings.Contains(outStdout, "imported state") {
		t.Fatalf("expected 'imported state' in dst stack output, got: %s", outStdout)
	}
}

// TestGetUpdateStatusViaHTTP tests the GET update status endpoint directly.
func TestGetUpdateStatusViaHTTP(t *testing.T) {
	truncateDB(t)
	env := newTestEnv(t, "update-status-project")
	env.login()

	fqn := devOrgLogin + "/update-status-project/dev"
	env.run("stack", "init", "--stack", fqn)
	env.run("up", "--yes", "--stack", fqn)

	// Get the latest update ID from the DB.
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	var updateID string
	err := dbPool.QueryRow(ctx, `SELECT id::text FROM updates ORDER BY created_at DESC LIMIT 1`).Scan(&updateID)
	if err != nil {
		t.Fatalf("query update ID: %v", err)
	}

	// Hit the get-update-status endpoint.
	resp := env.httpDo("GET", fmt.Sprintf("/api/stacks/%s/update-status-project/dev/update/%s", devOrgLogin, updateID), "")
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		t.Fatalf("expected 200, got %d", resp.StatusCode)
	}

	var results struct {
		Status string `json:"status"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&results); err != nil {
		t.Fatalf("decode response: %v", err)
	}
	if results.Status != "succeeded" {
		t.Fatalf("expected status 'succeeded', got %q", results.Status)
	}
}

File: e2e/state-operations.test.ts
// E2E — State operations: export empty stack, export after up, import roundtrip.

import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import path from "node:path";
import {
	apiRequest,
	BACKEND_URL,
	cleanupDir,
	createPulumiHome,
	newProjectDir,
	pulumi,
	truncateTables,
} from "./helpers.js";

const RANDOM_PET_PROGRAM = `name: state-proj
runtime: yaml
resources:
  pet:
    type: random:index:RandomPet
    properties:
      length: 2
outputs:
  petName: \${pet.id}
`;

describe("state operations", () => {
	let pulumiHome: string;
	let projectDir: string;
	let importUpdateId: string;

	beforeAll(async () => {
		pulumiHome = await createPulumiHome();
		await pulumi(["login", "--cloud-url", BACKEND_URL], { pulumiHome });
	});

	afterAll(async () => {
		if (projectDir) await cleanupDir(projectDir);
		await cleanupDir(pulumiHome);
		await truncateTables();
	});

	test("export empty stack returns valid deployment", async () => {
		await apiRequest("/stacks/dev-org/state-proj/empty", { method: "POST" });
		const res = await apiRequest("/stacks/dev-org/state-proj/empty/export");
		expect(res.status).toBe(200);
		const body = await res.json();
		expect(body).toHaveProperty("version");
		expect(body.version).toBe(3);
		expect(body).toHaveProperty("deployment");
	});

	test("export after pulumi up returns resources", async () => {
		projectDir = await newProjectDir("state-test");
		await Bun.write(path.join(projectDir, "Pulumi.yaml"), RANDOM_PET_PROGRAM);

		const initRes = await pulumi(["stack", "init", "dev-org/state-proj/with-resources"], {
			cwd: projectDir,
			pulumiHome,
		});
		expect(initRes.exitCode).toBe(0);

		const upRes = await pulumi(["up", "--yes"], { cwd: projectDir, pulumiHome });
		expect(upRes.exitCode).toBe(0);

		const exportRes = await apiRequest("/stacks/dev-org/state-proj/with-resources/export");
		expect(exportRes.status).toBe(200);
		const body = await exportRes.json();
		if (!body.deployment?.resources?.length) {
			const { SQL } = await import("bun");
			const { TEST_DB_URL } = await import("./helpers.js");
			const sql = new SQL({ url: TEST_DB_URL });
			const entries = await sql.unsafe(
				"SELECT kind, operation_id, sequence_id, remove_old, remove_new, (state IS NOT NULL) as has_state, (new_snapshot IS NOT NULL) as has_snap, elide_write FROM journal_entries ORDER BY sequence_id LIMIT 20",
			);
			console.error(`[debug] journal_entries (${entries.length} rows):`);
			for (const e of entries) {
				console.error(
					`  kind=${e.kind} opId=${e.operation_id} seq=${e.sequence_id} removeOld=${e.remove_old} removeNew=${e.remove_new} hasState=${e.has_state} hasSnap=${e.has_snap} elide=${e.elide_write}`,
				);
			}
			const cps = await sql.unsafe(
				"SELECT version, length(data::text) as bytes, created_at FROM checkpoints ORDER BY created_at DESC LIMIT 5",
			);
			console.error(`[debug] checkpoints (${cps.length} rows):`);
			for (const c of cps) {
				console.error(`  v=${c.version} bytes=${c.bytes} at=${c.created_at}`);
			}
			console.error(`[debug] deployment: ${JSON.stringify(body.deployment).slice(0, 500)}`);
			sql.close();
		}
		expect(body.deployment).toBeDefined();
		expect(body.deployment.resources).toBeArray();
		expect(body.deployment.resources.length).toBeGreaterThan(0);
	});

	test("export + import roundtrip", async () => {
		const exportRes = await apiRequest("/stacks/dev-org/state-proj/with-resources/export");
		const deployment = await exportRes.json();
		const resourceCount = deployment.deployment.resources.length;

		const importRes = await apiRequest("/stacks/dev-org/state-proj/with-resources/import", {
			method: "POST",
			body: deployment,
		});
		expect(importRes.status).toBe(200);
		const result = await importRes.json();
		expect(result).toHaveProperty("updateId");

		// Re-export and verify resources are intact
		const reExportRes = await apiRequest("/stacks/dev-org/state-proj/with-resources/export");
		const reBody = await reExportRes.json();
		expect(reBody.deployment.resources.length).toBe(resourceCount);
	});

	test("import to fresh stack", async () => {
		await apiRequest("/stacks/dev-org/state-proj/import-target", { method: "POST" });

		const exportRes = await apiRequest("/stacks/dev-org/state-proj/with-resources/export");
		const deployment = await exportRes.json();

		const importRes = await apiRequest("/stacks/dev-org/state-proj/import-target/import", {
			method: "POST",
			body: deployment,
		});
		expect(importRes.status).toBe(200);
		const result = await importRes.json();
		expect(result).toHaveProperty("updateId");
		importUpdateId = result.updateId;
	});

	test("get update status after import returns succeeded", async () => {
		const res = await apiRequest(
			`/stacks/dev-org/state-proj/import-target/update/${importUpdateId}`,
		);
		expect(res.status).toBe(200);
		const body = await res.json();
		expect(body.status).toBe("succeeded");
	});
});
